### PR TITLE
Adding initial CoprHD drive framework re #56

### PR DESCRIPTION
**Pull request by rodriguise**
This PR introduces the initial work for CoprHD support.

### CODE REVIEW

**Comment by rodriguise** [comment]
Ok should be good to go now.

**Comment by jvinod** [comment]
The files are missing a gofmt - if you use vim you can use https://github.com/fatih/vim-go.git

**Comment by rodriguise** [comment]
I use emacs, I added the hook and ran a gofmt -w on the files, let me know if it didn't take.

File: config.yaml [hunk 1]
```diff
@@ -4,11 +4,18 @@ osd:
     _nodeid: "1"
     _clusterid: "deadbeeef"
   drivers:
+    coprhd:
+     url: "https://172.16.70.20:4443/"
+     user: "root"
```
Comment (peter-edge) [inline config.yaml]: is this real?

File: volume/drivers/coprhd/coprhd.go [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+	"log"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/portworx/kvdb"
+
+	"github.com/libopenstorage/openstorage/api"
+	"github.com/libopenstorage/openstorage/volume"
+	"gopkg.in/jmcvetta/napping.v3"
```
Comment (peter-edge) [inline volume/drivers/coprhd/coprhd.go]: new dependencies must be vendored

File: volume/drivers/coprhd/coprhd.go [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+	consistency_group string
+	project           string
+	varray            string
+	vpool             string
+	url               string
+	user              string
+	password          string
+}
+
+type ApiError struct {
```
Comment (peter-edge) [inline volume/drivers/coprhd/coprhd.go]: s/Api/API/g

File: volume/drivers/coprhd/coprhd.go [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+	"gopkg.in/jmcvetta/napping.v3"
+)
+
+const (
+	Name = "coprhd"
+	Type = api.Block
+
+	// Place holders
+	RESTBaseUrl = "https://localhost:4443/"
+	APIUser     = "root"
```
Comment (peter-edge) [inline volume/drivers/coprhd/coprhd.go]: ? why are these defaults?

File: volume/drivers/coprhd/coprhd.go [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+	APIPassword = "ChangeMe"
+
+	// Common API URIs
+	LoginURI       = "login.json"
+	CreateVolURI   = "block/volumes.json"
+	DeleteVolURI   = "block/volumes/%s/deactivate.json"
```
Comment (peter-edge) [inline volume/drivers/coprhd/coprhd.go]: This is confusing calling this a URI, maybe URITemplate?

File: volume/drivers/coprhd/coprhd.go [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+
+func Init(params volume.DriverParams) (volume.VolumeDriver, error) {
+
+	consistency_group, ok := params["consistency_group"]
+
+	project, ok := params["project"]
+	if !ok {
+		project = "Default"
+	}
+
```
Comment (peter-edge) [inline volume/drivers/coprhd/coprhd.go]: nit: remove all new lines

File: volume/drivers/coprhd/coprhd.go [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+func (d *driver) String() string {
+	return Name
+}
+
+func (d *driver) Type() api.DriverType {
+	return Type
+}
+
+func init() {
```
Comment (peter-edge) [inline volume/drivers/coprhd/coprhd.go]: move to top

File: volume/drivers/coprhd/coprhd.go [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+	volume.Register(Name, Init)
+}
+
+// API volume create args
+type CreateVolumeArgs struct {
```
Comment (peter-edge) [inline volume/drivers/coprhd/coprhd.go]: GoDoc comments should match godoc style:

```
// CreateVolumeArgs is a ...
```

File: volume/drivers/coprhd/coprhd.go [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+	e := ApiError{}
+
+	token, err := d.getAuthToken()
+
+	if err != nil {
+		log.Printf("error: %s", err.Error())
+		return api.BadVolumeID, err
+	}
+
+	log.Printf("API auth token: %s\n\n", token)
```
Comment (peter-edge) [inline volume/drivers/coprhd/coprhd.go]: we are using logrus

Comment (peter-edge) [inline volume/drivers/coprhd/coprhd.go]: Do not log sensitive information

File: volume/drivers/coprhd/coprhd.go [hunk 10]
```diff
@@ -0,0 +1,271 @@
...
+			Transport: &http.Transport{
+				TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+			},
+		},
+		Header: &h,
+	}
+
+	res := CreateVolumeReply{}
+
+	sz := float64(spec.Size / (1024 * 1024 * 1000))
```
Comment (peter-edge) [inline volume/drivers/coprhd/coprhd.go]: 1014 * 1024 * 1000 should be a constant that is named appropriately

File: volume/drivers/coprhd/coprhd.go [hunk 11]
```diff
@@ -0,0 +1,271 @@
...
+		Name:              locator.Name,
+		Project:           d.project,
+		Size:              fmt.Sprintf("%.6fGB", sz),
+		Varray:            d.varray,
+		Vpool:             d.vpool,
+	}
+
+	resp, err := s.Post(url, &payload, &res, &e)
+
+	if resp.Status() == 202 {
```
Comment (peter-edge) [inline volume/drivers/coprhd/coprhd.go]: 202 as a constant (http.StatusAccepted?)

File: volume/drivers/coprhd/coprhd.go [hunk 12]
```diff
@@ -0,0 +1,271 @@
...
+
+		volId := res.Task[0].Resource.Id
+
```
Comment (peter-edge) [inline volume/drivers/coprhd/coprhd.go]: newlines

**Comment by gourao** [comment]
@rodriguise can you do a rebase. We'll merge this PR - just want to make sure it passes all the tests

**Comment by rodriguise** [comment]
Yeah will do in the AM

Sent from Outlook Mobile<https://aka.ms/qtex0l>




On Tue, Jan 12, 2016 at 5:56 PM -0800, "Gou Rao" <notifications@github.com<mailto:notifications@github.com>> wrote:


@rodriguise<https://github.com/rodriguise> can you do a rebase. We'll merge this PR - just want to make sure it passes all the tests

**Pull request closed without merging**